hedera-node/src/main/java/com/hedera/services/contracts/execution/traceability/SolidityAction.java

```diff
@@ -0,0 +1,179 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hedera.services.contracts.execution.traceability;
+
+import com.hedera.services.legacy.proto.utils.ByteStringUtils;
+import com.hedera.services.state.submerkle.EntityId;
+import com.hedera.services.stream.proto.ContractAction;
+
+public class SolidityAction {
+    private static final byte[] MISSING_BYTES = new byte[0];
+
+    private ContractActionType callType;
+    private final EntityId callingAccount;
+    private final EntityId callingContract;
+    private final long gas;
+    private final byte[] input;
+    private EntityId recipientAccount;
+    private EntityId recipientContract;
+    private byte[] invalidSolidityAddress;
+    private final long value;
+    private long gasUsed;
+    private byte[] output;
+    private byte[] revertReason;
+    private byte[] error;
+    private final int callDepth;
+
+    public SolidityAction(
+            final ContractActionType callType,
+            final EntityId callingAccount,
+            final EntityId callingContract,
+            final long gas,
+            final byte[] input,
+            final EntityId recipientAccount,
+            final EntityId recipientContract,
+            final long value,
+            final int callDepth) {
+        this.callType = callType;
+        this.callingAccount = callingAccount;
+        this.callingContract = callingContract;
+        this.gas = gas;
+        this.input = input == null ? MISSING_BYTES : input;
+        this.recipientAccount = recipientAccount;
+        this.recipientContract = recipientContract;
+        this.value = value;
+        this.callDepth = callDepth;
+    }
+
+    public ContractActionType getCallType() {
+        return callType;
+    }
+
+    public long getGas() {
+        return gas;
+    }
+
+    public byte[] getInput() {
+        return input;
+    }
+
+    public long getValue() {
+        return value;
+    }
+
+    public void setGasUsed(long gasUsed) {
+        this.gasUsed = gasUsed;
+    }
+
+    public byte[] getOutput() {
+        return output;
+    }
+
+    public void setOutput(final byte[] output) {
+        this.output = output;
+    }
+
+    public void setRevertReason(final byte[] revertReason) {
+        this.revertReason = revertReason;
+    }
+
+    public byte[] getError() {
+        return error;
+    }
+
+    public void setError(final byte[] error) {
+        this.error = error;
+    }
+
+    public void setInvalidSolidityAddress(byte[] invalidSolidityAddress) {
```

Comment:
Are these sets avoidable?  If so, could we use a record here instead?

Reply:
I'm not sure whether we can clean it up... new actions are created in an "incomplete" state and we can finalize them (record the `result`, `error`, `gasUsed`, etc) only after their associated frame has been completed.